Developer guide overhaul

tkknight opened the pull request:
## 🚀 Pull Request

### Description
<!-- Provide a clear description about your awesome pull request -->
<!-- Tell us all about your new feature, improvement, or bug fix -->

Overhaul of the **Developers Guide** section in the documentation.  Notable changes include:

- New sections:
  - Continuous Integration (CI) Testing
  - Getting Involved
  - Running the tests
- Updated sections:
  - Contributing (github convention) now points to the docs
  - Code Formatting
  - Graphics tests
  - Pull request check list
  - Reworking of the existing git guide  (gitwash)
  - Excluded some git urls from the linkcheck to avoid hitting the domain too many times and being blocked
  - PR template (points to new PR checklist)
- Various spelling and layout updates

For the lifetime of this PR you can view the rendered html here: 
- https://iris-test-doc.readthedocs.io/en/latest/

If you wish to read the changes as rendered, I suggest starting from the **Getting Involved** section and hitting **Next** to review the entire **Developers Guide** section: 
- https://iris-test-doc.readthedocs.io/en/latest/developers_guide/contributing_getting_involved.html

---

  1. [ ] **What's new** entry primed in a separate PR: https://github.com/SciTools/iris/pull/3854

---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/pulls.html#the-iris-check-list)

rcomer commented on docs/iris/src/developers_guide/contributing_documentation.rst:
```diff
@@ -1,49 +1,41 @@
 Building
---------
-
-The build is run from the documentation directory ``iris/docs/iris/src``.  In
```
I think it would be good to reinstate the info for what directory you need to be in.

rcomer reviewed (commented):
Hi @tkknight, thanks for doing this.  It's a massive improvement!  I don't think I've looked at everything, but just have some comments aimed at making it a bit more idiot-proof.

tkknight commented on docs/iris/src/developers_guide/contributing_pull_request_checklist.rst:
```diff
@@ -0,0 +1,62 @@
+=======================
+
+A pull request to a SciTools project master should be ready to merge into the
+master branch.
```
ok - I'm happy to remove the paragraph completely?

rcomer replied:
Sounds good to me 👍

tkknight commented on docs/iris/src/developers_guide/contributing_code_formatting.rst:
```diff
@@ -0,0 +1,59 @@
+
+or alternatively using ``conda``::
+
+    $ conda install -c conda-forge pre-commit
```
Indeed.  Will add a note.

tkknight commented:
Hi @rcomer.  I have addressed all your suggestions, the updated docs should be present in a few minutes (see it building: https://readthedocs.org/projects/iris-test-doc/builds/11893698/)

rcomer commented:
Thanks @tkknight, all the changes look good to me.  Think this still needs a review from somebody who understands all the technical ins and outs.  In that context, I'm a nobody!

trexfeathers commented on docs/iris/src/common_links.inc:
```diff
@@ -0,0 +1,24 @@
+.. _SciTools: https://github.com/SciTools
+.. _Iris: https://github.com/SciTools/iris
+.. _Iris GitHub: https://github.com/SciTools/iris
+.. _iris mailing list: https://groups.google.com/forum/#!forum/scitools-iris
```
Not something for this PR, but given its continued inclusion here, I'd just like to raise that checking the Google Group needs to be more enshrined as part of the role of a core dev. It's not something I check nearly enough.

tkknight replied:
I think the Google Groups usage is something we need to review - either start using it or make a statement of intent to stop using it.  I will capture this as an issue (discussion piece)

tkknight replied:
Discussion issue raised about google groups usage, see https://github.com/SciTools/iris/issues/3860

trexfeathers reviewed (changes requested):
First of all, @tkknight and @bjlittle I just want to say what a excellent job you've done here 😍

I'm sure I'm not the only one who has in the past been cautioned against following the existing guide, for fear of it being outdated. But now I can see all the necessary knowledge - which many of us have learned in patches, with some difficulty - all combined into a single easy to follow guide - brilliant!

I've found a few minor points of grammar and readability that could be improved. Since @tkknight is enjoying some **well earned** time off I'm sure I can find someone else on the release team to push changes to this branch.

trexfeathers commented on docs/iris/src/developers_guide/contributing_documentation.rst:
```diff
@@ -113,36 +113,36 @@ or ignore the url.
+custom set of python scripts are used, these are located in the directory
+``iris/docs/iris/src/sphinxext``.  Once the ``make html`` command has been run,
 the output of these scripts can be found in
 ``iris/docs/iris/src/_build/generated/api``.
```
07cc7eda980ce2ac1e37007c267c35e0e565c9e1

trexfeathers commented on docs/iris/src/developers_guide/contributing_code_formatting.rst:
```diff
@@ -0,0 +1,63 @@
+         be present.
+
+In order to to install the ``pre-commit`` git hooks defined in our
+``.pre-commit-config.yaml`` file you must run the following command from the
```
232f384223ab69017e61454bdb72d0f6aa41fa63

trexfeathers commented on docs/iris/src/developers_guide/documenting/docstrings.rst:
```diff
@@ -4,39 +4,58 @@
+* ``u"""Unicode triple-quoted string"""`` for Unicode docstrings
 
-All docstrings should be written in rST (reStructuredText) markup; an rST guide follows this page.
+All docstrings should be written in rST (reStructuredText) markup.  See the
```
232f384223ab69017e61454bdb72d0f6aa41fa63

trexfeathers commented on docs/iris/src/developers_guide/documenting/rest_guide.rst:
```diff
@@ -1,7 +1,15 @@
 reST quick start
 ================
 
+reST (http://en.wikipedia.org/wiki/ReStructuredText) is used to create the
```
232f384223ab69017e61454bdb72d0f6aa41fa63

trexfeathers commented on docs/iris/src/developers_guide/contributing_graphics_tests.rst:
```diff
@@ -0,0 +1,166 @@
+Graphics tests
+**************
+
+Iris may be used Iris to create various forms of graphical output, to ensure
```
232f384223ab69017e61454bdb72d0f6aa41fa63

trexfeathers commented on docs/iris/src/developers_guide/contributing_graphics_tests.rst:
```diff
@@ -0,0 +1,166 @@
+The `Iris Travis matrix`_ defines multiple test runs that use
+different versions of Python to ensure Iris is working as expected.
+
+To make this manageable, the :meth:`iris.tests.IrisTest.check_graphic` test
```
232f384223ab69017e61454bdb72d0f6aa41fa63

trexfeathers commented on docs/iris/src/developers_guide/contributing_graphics_tests.rst:
```diff
@@ -0,0 +1,166 @@
+
+This consists of:
+
+ * The :meth:`iris.tests.IrisTest.check_graphic` function uses a perceptual **image hash** of the outputs
```
232f384223ab69017e61454bdb72d0f6aa41fa63

trexfeathers commented on docs/iris/src/developers_guide/contributing_ci_tests.rst:
```diff
@@ -0,0 +1,62 @@
+be performed to ensure that all expected variations test successfully.
+
+The `travis-ci`_ tests are run automatically against the `Iris`_ master
+repository when a pull requests is submitted, updated or merged.
```
232f384223ab69017e61454bdb72d0f6aa41fa63

trexfeathers commented on docs/iris/src/developers_guide/contributing_ci_tests.rst:
```diff
@@ -0,0 +1,62 @@
+
+.. image:: ci_checks.png
+
+If any checks fail the pull requests is unlikely to be merged to the Iris
```
232f384223ab69017e61454bdb72d0f6aa41fa63

trexfeathers commented on docs/iris/src/developers_guide/contributing_ci_tests.rst:
```diff
@@ -0,0 +1,62 @@
+The `travis-ci`_ tests are run automatically against the `Iris`_ master
+repository when a pull requests is submitted, updated or merged.
+
+The example snapshot from a GitHub pull request shows all test passing.
```
232f384223ab69017e61454bdb72d0f6aa41fa63

trexfeathers commented on docs/iris/src/developers_guide/contributing_ci_tests.rst:
```diff
@@ -0,0 +1,62 @@
+
+A `travis-ci`_ configuration file named `.travis.yml`_
+is in the Iris repository which tells travis-ci what commands to run.  The
+commands includes retrieving the Iris code base and associated test files using
```
232f384223ab69017e61454bdb72d0f6aa41fa63

bjlittle commented:
@trexfeathers Thanks for the pair review to see this over the line... over to you now.

trexfeathers commented:
🎉🎉

rcomer commented:
Awesome work y'all!
